engines/e_pkcs11_eng.c

```diff
@@ -0,0 +1,486 @@
+    if (len < 0)
+        return 0;
+
+    *out = der;
+    *out_len = len;
+    return 1;
+}
+
+static char pkcs11_hex_int(char nib1, char nib2)
+{
+    int ret = (nib1-(nib1 <= 57 ? 48 : (nib1 < 97 ? 55 : 87)))*16;
+    ret += (nib2-(nib2 <= 57 ? 48 : (nib2 < 97 ? 55 : 87)));
+    return ret;
+}
+
+static char* pkcs11_hex2a(char *hex)
+{
+    int vlen, j = 0, i, ishex;
+    char *hex2a;
+
+    hex2a = OPENSSL_malloc(strlen(hex) + 1);
+
+    if (hex2a == NULL) {
+        return NULL;
+    }
+
+    vlen = strlen(hex);
+    ishex = pkcs11_ishex(hex);
+    for (i = 0; i < vlen; i++) {
+        if ((*(hex+i) == '%' && i < (vlen-2)) || ishex) {
+            *(hex2a+j) = pkcs11_hex_int(*(hex+i+1-ishex), *(hex+i+2-ishex));
+            i += (2-ishex);
+        } else {
+            *(hex2a+j) = *(hex+i);
+        }
+        j++;
+    }
+    *(hex2a+j) = '\0';
+    return hex2a;
+}
+
+static int pkcs11_ishex(char *hex)
+{
+    int i,h = 0;
+    size_t len;
+
+    len = strlen(hex);
+    for (i = 0; i < len; i++) {
+        if ((*(hex) >= '0' && *(hex) <= '9') || (*(hex) >= 'a' && *(hex) <= 'f')
+            || (*(hex) >= 'A' && *(hex) <= 'F')) h++;
+    }
+    if (h == len && !(h % 2))
+        return 1;
+    return 0;
+}
+
+static int pkcs11_parse_uri(const char *path, char *token, char **value)
+{
+    char *tmp, *end, *hex2a;
+    size_t tmplen;
+
+    if ((tmp = strstr(path, token)) == NULL)
+        return 0;
+    tmp += strlen(token);
+    tmplen = strlen(tmp);
+    *value = OPENSSL_malloc(tmplen + 1);
+
+    if (*value == NULL) {
+        PKCS11err(PKCS11_F_PKCS11_PARSE_URI, ERR_R_MALLOC_FAILURE);
+        goto err;
+    }
+
+    end = strpbrk(tmp, ";");
+    BIO_snprintf(*value, end == NULL ? tmplen + 1 :
+                 (size_t) (end - tmp + 1), "%s", tmp);
+
+    hex2a = pkcs11_hex2a(*value);
+    free(*value);
+    *value = hex2a;
+    return 1;
+
+ err:
+    free(*value);
+    *value = NULL;
+    return 0;
+}
+
+static int pkcs11_get_console_pin(char **pin)
+{
+#ifndef OPENSSL_NO_UI_CONSOLE
+    int i;
+    const int buflen = 512;
+    char *strbuf = OPENSSL_malloc(buflen);
+
+    if (strbuf == NULL) {
+        PKCS11err(PKCS11_F_PKCS11_GET_CONSOLE_PIN, ERR_R_MALLOC_FAILURE);
+        return 0;
+    }
+
+    for (;;) {
+        char prompt[200];
+        BIO_snprintf(prompt, sizeof(prompt), "Enter PIN: ");
+        strbuf[0] = '\0';
+        i = EVP_read_pw_string((char *)strbuf, 512, prompt, 1);
+        if (i == 0) {
+            if (strbuf[0] == '\0') {
+                goto err;
+            }
+            *pin = strbuf;
+            return 1;
+        }
+        if (i < 0) {
+            PKCS11_trace("bad password read\n");
+            goto err;
+        }
+    }
+
+ err:
+    OPENSSL_free(strbuf);
+#endif
+
+    return 0;
+}
+
+static int pkcs11_parse(PKCS11_CTX *ctx, const char *path)
+{
+    char *id, *module_path = NULL;
+    char *pin = NULL, *label = NULL, *slotid = NULL;
+
+    ctx->slotid = 0;
```

Comment:
better write `=NULL`, not `= 0`

Reply:
ctx->slotid is ulong

Reply:
My bad, sorry  😕 
I make a larger review to compensate ;)